crate/server/src/core/implementation.rs

```diff
@@ -50,31 +50,44 @@ use crate::{
-            ),
-            DbParams::Sqlite(db_path) => Box::new(
-                SqlitePool::instantiate(&db_path.join("kms.db"), shared_config.clear_db_on_start)
+    pub async fn instantiate(mut shared_config: ServerParams) -> KResult<Self> {
+        let db: Box<dyn Database + Sync + Send> = if let Some(mut db_params) =
+            shared_config.db_params.as_mut()
+        {
+            match &mut db_params {
+                DbParams::SqliteEnc(db_path) => Box::new(
+                    CachedSqlCipher::instantiate(db_path, shared_config.clear_db_on_start).await?,
+                ),
+                DbParams::Sqlite(db_path) => Box::new(
+                    SqlitePool::instantiate(
+                        &db_path.join("kms.db"),
+                        shared_config.clear_db_on_start,
+                    )
                     .await?,
-            ),
-            DbParams::Postgres(url) => {
-                Box::new(PgPool::instantiate(url, shared_config.clear_db_on_start).await?)
-            }
-            DbParams::Mysql(url) => {
-                Box::new(MySqlPool::instantiate(url, shared_config.clear_db_on_start).await?)
-            }
-            DbParams::RedisFindex(url, master_key, label) => {
-                // There is no reason to keep a copy of the key in the shared config
-                // So we are going to create a "zeroizable" copy which will be passed to Redis with Findex
-                // and zerorize the one in the shared config
-                let key_bytes = master_key.to_bytes();
-                let new_master_key =
-                    SymmetricKey::<REDIS_WITH_FINDEX_MASTER_KEY_LENGTH>::try_from_bytes(key_bytes)?;
-                master_key.zeroize();
-                Box::new(RedisWithFindex::instantiate(url, new_master_key, label).await?)
+                ),
+                DbParams::Postgres(url) => Box::new(
+                    PgPool::instantiate(url.as_str(), shared_config.clear_db_on_start).await?,
+                ),
+                DbParams::Mysql(url) => Box::new(
+                    MySqlPool::instantiate(url.as_str(), shared_config.clear_db_on_start).await?,
+                ),
+                DbParams::RedisFindex(url, master_key, label) => {
+                    // There is no reason to keep a copy of the key in the shared config
+                    // So we are going to create a "zeroizable" copy which will be passed to Redis with Findex
+                    // and zerorize the one in the shared config
+                    let key_bytes = master_key.to_bytes();
+                    let new_master_key =
+                        SymmetricKey::<REDIS_WITH_FINDEX_MASTER_KEY_LENGTH>::try_from_bytes(
+                            key_bytes,
+                        )?;
+                    master_key.zeroize();
+                    Box::new(
+                        RedisWithFindex::instantiate(url.as_str(), new_master_key, label).await?,
+                    )
```

Comment:
- wrt to `to_bytes()` : I do not think it is a problem since it does not go into the heap and stays on the stack; I guess the compiler did this for me but jut in case, I rewrote this :
```Rust
let new_master_key =
                        SymmetricKey::<REDIS_WITH_FINDEX_MASTER_KEY_LENGTH>::try_from_bytes(
                            master_key.to_bytes(),
                        )?;
```
(SymmetricKey will go to the heap but implements zeroize)

- master_key needs to be zeroized: it is a (mutable) field of a struct (ServerParams) which is shared between threads and stays on the heap for the duration of the program (~static)

Reply:
The original master key is placed in the same object (which type is `KMS`) as the `db` object. Therefore it seems both the old master key and the `db` objects are handled with the same precautions. If that's true,  then I see no reason to zeroize the old master key as it will be kept (as) private (as the `db`). Is there another reason for zeroizing it? Not doing so makes this match arm a one-liner and removes the needs for mutability of the `db_params`.

As for the stack vs heap allocations and zeroization, it seems that [stack-allocated structures are still an issue for `Copy` and moves](https://docs.rs/zeroize/latest/zeroize/#stackheap-zeroing-notes). But since we are inside the KMS, this should not matter.

```rust
        let db: Box<dyn Database + Sync + Send> = if let Some(db_params) = &shared_config.db_params
        {
            match db_params {
                DbParams::SqliteEnc(db_path) => Box::new(
                    CachedSqlCipher::instantiate(db_path, shared_config.clear_db_on_start).await?,
                ),
                DbParams::Sqlite(db_path) => Box::new(
                    SqlitePool::instantiate(
                        &db_path.join("kms.db"),
                        shared_config.clear_db_on_start,
                    )
                    .await?,
                ),
                DbParams::Postgres(url) => Box::new(
                    PgPool::instantiate(url.as_str(), shared_config.clear_db_on_start).await?,
                ),
                DbParams::Mysql(url) => Box::new(
                    MySqlPool::instantiate(url.as_str(), shared_config.clear_db_on_start).await?,
                ),
                DbParams::RedisFindex(url, master_key, label) => {
                    Box::new(RedisWithFindex::instantiate(url.as_str(), master_key, label).await?)
                }
            }
        } else {
            kms_bail!("Fatal: no database configuration provided. Stopping.")
        };
```

Reply:
OK. Here is what I propose. Let us fix this in crypto-core: returning `Zerozing<[u8;LENGTH>` rather than just the array for secrets will avoid having to wonder what happens down the line and fixes the stack memory issue you mentioned earlier.